folio-token/src/main/java/org/folio/spring/service/SystemUserService.java

```diff
@@ -0,0 +1,146 @@
+    if (systemUserCache == null) {
+      return getSystemUser(tenantId);
+    }
+
+    var user = systemUserCache.get(tenantId, this::getSystemUser);
+    var userToken = user.token();
+    var now = Instant.now();
+    if (userToken.accessTokenExpiration().isAfter(now)) {
+      return user;
+    }
+
+    systemUserCache.invalidate(tenantId);
+    user = getSystemUser(tenantId);
+    systemUserCache.put(tenantId, user);
+
+    return user;
+  }
+
+  /**
+   * Authenticate system user and return token value.
+   *
+   * @param user {@link SystemUser} to log with
+   * @return token value
+   */
+  public UserToken authSystemUser(SystemUser user) {
+    var token = getTokenWithExpiry(user);
+    if (isNull(token)) {
+      token  = getTokenLegacy(user);
+    }
+    return token;
+  }
+
+  @Autowired(required = false)
+  public void setSystemUserCache(Cache<String, SystemUser> systemUserCache) {
+    this.systemUserCache = systemUserCache;
+  }
+
+  private SystemUser getSystemUser(String tenantId) {
+    log.info("Attempting to issue token for system user [tenantId={}]", tenantId);
+    var systemUser = SystemUser.builder()
+        .tenantId(tenantId)
+        .username(systemUserProperties.username())
+        .okapiUrl(environment.getOkapiUrl())
+        .build();
+
+    // create context for authentication
+    try (var fex = new FolioExecutionContextSetter(contextBuilder.forSystemUser(systemUser))) {
+      var token = authSystemUser(systemUser);
+      systemUser = systemUser.withToken(token);
+      log.info("Token for system user has been issued [tenantId={}]", tenantId);
+    }
+    // create context for user with token for getting user id
+    try (var fex = new FolioExecutionContextSetter(contextBuilder.forSystemUser(systemUser))) {
+      var userId = prepareUserService.getFolioUser(systemUserProperties.username())
+          .map(UsersClient.User::id).orElse(null);
+      return systemUser.withUserId(userId);
+    }
+  }
+
+  private UserToken getTokenLegacy(SystemUser user) {
+    var response =
+        authnClient.login(new UserCredentials(user.username(), systemUserProperties.password()));
+
+    if (!isNull(response) && response.getStatusCode() == HttpStatusCode.valueOf(404)) {
+      return null;
+    }
+    var accessToken =  ofNullable(response.getHeaders()
+        .get(X_OKAPI_TOKEN))
+        .orElseThrow(() -> new AuthorizationException("Cannot retrieve okapi token for tenant: " + user.username()))
+        .get(0);
+
+    return UserToken.builder()
+        .accessToken(accessToken)
+        .accessTokenExpiration(Instant.MAX)
+        .build();
+  }
+
+  private UserToken getTokenWithExpiry(SystemUser user) {
+    var response =
+        authnClient.loginWithExpiry(new UserCredentials(user.username(), systemUserProperties.password()));
+
+    if (!isNull(response) && response.getStatusCode() == HttpStatusCode.valueOf(404)) {
+      return null;
+    }
+    if (!isNull(response) && isNull(response.getBody())) {
```

Comment:
This should fix a bug found by Sonar.
Also It would be good to move this check up so we don't need a response null check in another 'if'
```suggestion
    if (isNull(response) || isNull(response.getBody())) {
```